File: tape/sdk/typescript/src/connectors/pubsub.ts
import type {
  Connector, EffectRecord, ObligationRecord,
  DispatchResult, ObservationResult, CompensationResult,
} from './index.ts';

export interface PubSubConnectorOpts {
  name?: string;
  project: string;
  topic: string;
  compensateTopic?: string;
  /** Optional Tape URL used by Observe() to look up the subscriber's result record. */
  tapeUrl?: string;
}

/**
 * Publish the intent as a Pub/Sub message. The subscriber writes a result
 * record back to Tape (`tape.setValue("outbox/<connector>", key, {...})`)
 * which Observe() reads to resolve UNKNOWN. `@google-cloud/pubsub` is an
 * optional dependency — lazy-imported.
 */
export class PubSubConnector implements Connector {
  readonly name: string;
  private readonly opts: PubSubConnectorOpts;
  private client: any = null;

  constructor(opts: PubSubConnectorOpts) {
    if (!opts.project || !opts.topic) {
      throw new Error('PubSubConnector: project and topic are required');
    }
    this.name = opts.name ?? `pubsub:${opts.topic}`;
    this.opts = opts;
  }

  private async ensure(): Promise<any> {
    if (this.client) return this.client;
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const mod = await import('@google-cloud/pubsub' as any);
      const PubSub = (mod as any).PubSub;
      this.client = new PubSub({ projectId: this.opts.project });
      return this.client;
    } catch (ex) {
      throw new Error(
        'PubSubConnector requires @google-cloud/pubsub — `npm i @google-cloud/pubsub`. ' +
        `Underlying: ${(ex as Error).message}`,
      );
    }
  }

  async dispatch(effect: EffectRecord): Promise<DispatchResult> {
    try {
      const c = await this.ensure();
      const topic = c.topic(this.opts.topic, { messageOrdering: true });
      const messageId = await topic.publishMessage({
        json: effect.payload,
        orderingKey: effect.runId,
        attributes: {
          tape_idempotency_key: effect.idempotencyKey,
          tape_run_id: effect.runId,
          tape_business_key: effect.businessKey ?? '',
          tape_tool: effect.toolName,
          tape_attempt: String(effect.attempt ?? 1),
        },
      });
      return { outcome: 'confirmed', response: { message_id: messageId }, dispatchId: messageId };
    } catch (ex) {
      return { outcome: 'unknown', error: (ex as Error).message };
    }
  }

  async observe(_effect: EffectRecord): Promise<ObservationResult> {
    // The Pub/Sub topology resolves UNKNOWN via the subscriber writing a
    // tape value the reconciler reads. To keep this file dep-light, we
    // surface a documented hook: chain an `HttpConnector.observeUrl`, or
    // override this method in a subclass.
    return { outcome: 'unknown', error: 'PubSubConnector.observe: chain a status URL via HttpConnector.observeUrl, or subclass.' };
  }

  async compensate(obligation: ObligationRecord): Promise<CompensationResult> {
    if (!this.opts.compensateTopic) {
      return { outcome: 'stuck', error: 'no compensateTopic configured' };
    }
    try {
      const c = await this.ensure();
      const topic = c.topic(this.opts.compensateTopic, { messageOrdering: true });
      const messageId = await topic.publishMessage({
        json: obligation.payload,
        orderingKey: obligation.runId,
        attributes: {
          tape_obligation_kind: obligation.kind,
          tape_effect_key: obligation.effectKey,
          tape_run_id: obligation.runId,
        },
      });
      return { outcome: 'compensated', response: { message_id: messageId } };
    } catch (ex) {
      return { outcome: 'pending', error: (ex as Error).message };
    }
  }
}

File: tape/sdk/go/connectors/pubsub.go
// Pub/Sub connector — publish the intent as a Pub/Sub message.
//
// The real google-cloud-pubsub client (`cloud.google.com/go/pubsub`) is an
// optional dependency. To keep the SDK build light, this file ships in two
// halves: the always-compiled stub (this file) which reports "not built"
// errors, and a `pubsub_real.go` that's compiled with `-tags pubsub`. This
// mirrors the Pub/Sub bridge in `reactions_pubsub.go`.

//go:build !pubsub

package connectors

import (
	"context"
	"errors"
)

// PubSubOpts — configuration. With the default build, these values are
// preserved on the connector but `Dispatch` returns ErrPubSubNotBuilt.
type PubSubOpts struct {
	Name            string
	Project         string
	Topic           string
	CompensateTopic string
	TapeURL         string // for the Observe() path via tape.GetValue
}

// ErrPubSubNotBuilt — returned by the default build of the Pub/Sub
// connector. Build with `-tags pubsub` to enable, and `go get
// cloud.google.com/go/pubsub`.
var ErrPubSubNotBuilt = errors.New("connectors/pubsub: build with -tags pubsub")

// PubSubConnector — Pub/Sub-backed outbox connector.
type PubSubConnector struct {
	opts PubSubOpts
}

// NewPubSubConnector — construct (Dispatch returns ErrPubSubNotBuilt in
// the default build).
func NewPubSubConnector(opts PubSubOpts) *PubSubConnector {
	if opts.Name == "" {
		opts.Name = "pubsub:" + opts.Topic
	}
	return &PubSubConnector{opts: opts}
}

func (c *PubSubConnector) Name() string { return c.opts.Name }

func (c *PubSubConnector) Dispatch(ctx context.Context, e Effect) (DispatchResult, error) {
	return DispatchResult{Outcome: DispatchUnknown, Error: ErrPubSubNotBuilt.Error()}, ErrPubSubNotBuilt
}

func (c *PubSubConnector) Observe(ctx context.Context, e Effect) (ObservationResult, error) {
	return ObservationResult{Outcome: ObservationUnknown, Error: ErrPubSubNotBuilt.Error()}, ErrPubSubNotBuilt
}

func (c *PubSubConnector) Compensate(ctx context.Context, o Obligation) (CompensationResult, error) {
	return CompensationResult{Outcome: CompensationStuck, Error: ErrPubSubNotBuilt.Error()}, ErrPubSubNotBuilt
}
